### DISCUSSION TITLE
Tokens & pocketbase restart

### DISCUSSION THREAD

**Discussion by vkuznecovas** [pocketbase/pocketbase#911] [Q&A] [answered]
Sorry if this has been asked before but I could not find an answer.

To my understanding, whenever pocketbase is restarted the existing tokens for users become invalid. 

My setup involves running pocketbase as a framework with `v0.8.0-rc2` version as well as a web app using the pocketbase js SDK(`@next`). I have a collection with an update rule of `@request.auth.id=user`. The user field in the collection is a relation to the `users` auth collection.

I'm saying this because I've been noticing this behaviour:

1) I login as a user.
2) I update an entry in the collection successfully via the sdk.
3) I restart my pocketbase server.
4) Using the same token, the same request now results in the following response:
```
{
    "code": 404,
    "message": "The requested resource wasn't found.",
    "data": {}
}
```

Clearing the local storage and logging in again solves the issue.

I've noticed that the token generation involves a key that is generated by taking the `record.TokenKey()` and adding `app.Settings().RecordAuthToken.Secret` to it. The `app.Settings().RecordAuthToken` is regenerated on startup of pocketbase and I assume this is the culprit.

1) Would it be feasible to make it so that we can pass in an ENV var to "freeze" this setting instead of regenerating this at run time? Are there any security concerns with this approach?
2) If 1) is not an option, what would be the recommended way to deal with this client side?

**Comment by ganigeorgiev** [comment]
> To my understanding, whenever pocketbase is restarted the existing tokens for users become invalid.

That's not correct. Tokens are invalidated only when:

- the admin explicitly invalidate the base token keys from "Admin UI > Settings > Token Options"
- the user/admin change their password

If that wasn't the case, the integration tests wouldn't pass.

~~Your issue is probably client-side related. Could you provide a reproducible code sample or some context how do you use the SDK (eg. is it in a SSR server action or executed in the browser, etc.)?~~

**Edit: This is not correct. See the updated answer in the comment below https://github.com/pocketbase/pocketbase/discussions/911#discussioncomment-4045608.**

**Reply by vkuznecovas** [reply]
So here's a curl example, I've omitted the bodies for brevity:

Pre pocketbase restart, after logging in and getting a token.
```
curl --location --request PATCH 'http://localhost:8090/api/collections/listed_item/records/8bqw7bzyqz8ukh0' \
--header 'Authorization: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb2xsZWN0aW9uSWQiOiJzeXN0ZW1wcm9maWxlczAiLCJleHAiOjE2Njg2NzQ5NDQsImlkIjoiaGJjd2pteXV6MXpyZmJsIiwidHlwZSI6ImF1dGhSZWNvcmQifQ.kjF6cCYKUtBiq_0CQGp_a8nwZF4V3jvwbq4KplrLVZ8' \
--header 'Content-Type: application/json' \
--data-raw '{
    "accuracy": 56.5
    ...
}'
{
  "accuracy": 56.5
  ...
}
```

Post pocketbase restart, using the same token:
```
curl --location --request PATCH 'http://localhost:8090/api/collections/listed_item/records/8bqw7bzyqz8ukh0' \
--header 'Authorization: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb2xsZWN0aW9uSWQiOiJzeXN0ZW1wcm9maWxlczAiLCJleHAiOjE2Njg2NzQ5NDQsImlkIjoiaGJjd2pteXV6MXpyZmJsIiwidHlwZSI6ImF1dGhSZWNvcmQifQ.kjF6cCYKUtBiq_0CQGp_a8nwZF4V3jvwbq4KplrLVZ8' \
--header 'Content-Type: application/json' \
--data-raw '{
    "accuracy": 56.5,
    ...
}'
{
  "code": 404,
  "message": "The requested resource wasn't found.",
  "data": {}
}
```

After re-logging in, with new token:
```
curl --location --request PATCH 'http://localhost:8090/api/collections/listed_item/records/8bqw7bzyqz8ukh0' \
--header 'Authorization: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb2xsZWN0aW9uSWQiOiJzeXN0ZW1wcm9maWxlczAiLCJleHAiOjE2Njg2ODAxMTMsImlkIjoiaGJjd2pteXV6MXpyZmJsIiwidHlwZSI6ImF1dGhSZWNvcmQifQ.0kULkgQGwZ908WKlh_5Y-4Bdm4n3ghmxYRqRs0JpJGw' \
--header 'Content-Type: application/json' \
--data-raw '{
    "accuracy": 56.5,
    ...
}'
{
  "accuracy": 56.5,
  ...
}
```

I am doing some unorthodox things with pocketbase though, but it does not seem to be related to that.

**Comment by ganigeorgiev** [comment, marked as answer]
@vkuznecovas Actually, I stand corrected. The issue you are experiencing is possible if you have migrated from v0.7 -> v0.8.

To fix it, go to "Admin UI > Settings" and make any settings change (for example updating the app name).

The reason why this will fix it is because in v0.8 the settings user token fields are renamed but the upgrade command doesn't persist it and basically when you restart PocketBase you have always the old settings fields merged with the default values for the new ones (which are some random autogenerated).

For new v0.8.0 projects the above is not an issue because we perform an explicit save when you first initialize the application and there are no other settings - https://github.com/pocketbase/pocketbase/blob/rc/core/base.go#L361-L364

I'll make sure to update the `upgrade` command and/or add settings change detection in the app initialization.

**Reply by vkuznecovas** [reply]
Thanks! This has fixed the issue.
